consensus/consensus.go

```diff
@@ -85,8 +83,58 @@ func (fc *FastHotStuffConsensus) Init(
 	return nil
 }
 
-func (fc *FastHotStuffConsensus) UpdateChainTip( /* TODO */ ) {
-	// TODO
+// UpdateChainTip is called when the caller has connected a new block at the tip of the blockchain,
+// for an arbitrary block height and view. Given the new block has been connected, this function
+// resets the internal timers and state of the Fast HotStuff consensus that determine the next action.
+// The functions expects the following for the input params:
+//   - chainTip: the input block must have a valid block hash, block height, view, and QC
+//   - validators: the validators must be sorted in decreasing order of stake, with a
+//     consistent tie breaking scheme. The validator set is expected to be valid for
+//     validating votes and timeouts for the next block height.
+func (fc *FastHotStuffConsensus) UpdateChainTip(chainTip Block, validators []Validator) error {
```

Comment:
I've been thinking about this function for some time, and I'm a little worried about the assumptions made here. Given that we're calling this function in server's `_handleBlockAccepted`, we assume that `_handleBlockAccepted` is called with blocks with monotonically increasing views (note that if it isn't then `evictStaleVotesAndTimeouts` could corrupt our votes/timeouts). This is a strong assumption, and if it is true, then server's function should really be `_handleBlockCommitted` rather than "accepted". But then fasthotstuff's function should be called `ProcessCommittedBlock` or something like this.

 I feel like with this function, we are deviating from the pattern in the reference implementation. We are advancing the view without explicitly validating a QC in the block, that's a little tricky. Overall, I think this function is making a lot of implicit assumptions that could have complex ramifications depending on how the other code fits together. I would try to get as close to the `handleBlockFromPeer` pattern that we used in the [FastHotstuff reference](https://github.com/deso-protocol/hotstuff_pseudocode/pull/7). Maybe you can put my mind at ease if you elaborate on how `UpdateChainTip` and `_handleBlockAccepted` get called precisely. Otherwise, I think this is a dangerous public function and we should either make it private or rewrite it.